File: beecrowd/beecrowd_1210.py
"""
  Produção Ótima de Ótima Vodka #1210

  Link: https://www.beecrowd.com.br/judge/pt/problems/view/1210

  Author: Elias E. S. Rodrigues
"""

def optimal_production_vodka(periodo, idade, maxima, preco, custo, venda, custos, trocas):
  """
    Calcula todas as possibilidades de troca ou manutenção durante o período
    proposto e a idade do destilador de 0 até máxima.

    Para cada (período, idade) verifica se é melhor trocar ou dar manutenção:
      trocar = preço do novo + custo da manutenção - valor da venda + custo acumulado
      manutenção = custo da manutenção + custo acumulado

    Se o valor de trocar for menor ou igual ao valor da manutenção, então é
    recomendado trocar. É adicionado na lista de custos e trocas, os valores:
      custos[período, idade] = valor de trocar
      trocas[período, idade] = True, pq foi trocado
    Caso contrário é recomendado dar manutenção, então é armazenado o
    valor acumulado da manutenção na lista custos.
      custos[(período, idade)] = valor da manuteção
  """
  for i in range(periodo, 0, -1):
    for j in range(maxima+1):
      troca = preco + custo[0] - venda[j-1] + custos[i+1][1]
      manutencao = custo[j] + custos[i+1][j+1]

      if troca <= manutencao:
        custos[i][j] = troca
        trocas.append((i, j))
      else:
        custos[i][j] = manutencao

    """
      Quando o destilador chega a idade máxima, então tem que trocar
      obrigatoriamente.
    """
    troca = preco + custo[0] - venda[j-1] + custos[i+1][1]
    custos[i][j] = troca
    trocas.append((i, j))

  """
    Verifica os valores dos períodos onde houve troca e monta
    a saída de anos. Os anos de troca são as posições [período, idade],
    onde o período e a idade do destilador foram trocados e que estão
    dentro do período proposto.
  """
  idade_i = idade
  anos = []
  for i in range(1, periodo+1):
    if (i, idade_i) in trocas:
      anos.append(str(i))
      anos.append(' ')
      idade_i = 1
    else:
      idade_i += 1

  anos = ''.join(anos[0:-1]) if len(anos) != 0 else ['0', ' ']

  return custos[1][idade], anos

#   return custos[1][idade], ''.join(anos[0:-1])

if __name__ == '__main__':
  
  """ Arquivo de teste. """
  # file = open('./beecrowd_1210_teste.txt', 'r')
  # inputs = file.readlines()
  # file.close()
  # f = 0
  # file = open('./beecrowd_1210_resultado.txt', 'w')

  custos = [[0 for i in range(2001)] for j in range(2001)]

  while True:
    try:
      """ Arquivo de teste. """
      # periodo, idadeI, idadeM, preco = inputs[f].split()
      # f += 1
      # dados = inputs[f].split()
      # custo = [int(i) for i in dados]
      # custo += [0]
      # f += 1
      # dados = inputs[f].split()
      # venda = [int(i) for i in dados]
      # venda += [0]
      # f += 1

      periodo, idadeI, idadeM, preco = input().split()
      dados = input().split()
      custo = [int(i) for i in dados]
      custo.append(0)
      dados = input().split()
      venda = [int(i) for i in dados]
      venda.append(0)

      trocas = []

      total, anos = optimal_production_vodka(int(periodo), int(idadeI), int(idadeM), int(preco), custo, venda, custos, trocas)
      
      print(f'{total}\n{anos}')

      # file.write(f'{custos[1][idade]}\n{anos}\n')

    except EOFError:
      break

  # file.close()

File: beecrowd/beecrowd_1210.c
/*
  Produção Ótima de Ótima Vodka #1210

  Link: https://www.beecrowd.com.br/judge/pt/problems/view/1210

  Author: Elias E. S. Rodrigues
*/

#include <stdio.h>
#include <stdlib.h>

int main(void) {
  int periodo, idade, maxima, preco, custo[2000], venda[2001];
  int **custos, **trocas, i, j, troca, manutencao, espaco;

  while (scanf("%d %d %d %d", &periodo, &idade, &maxima, &preco) != EOF) {
    /*
      Pega os vetores de custo de manutencao e venda
      e aloca dinamicamente os vetores de custos acumulados
      e trocas.
    */
    for (i = 0; i < maxima; i++) {
      scanf("%d", &custo[i]);
    }
    for (i = 1; i <= maxima; i++) {
      scanf("%d", &venda[i]);
    }
    custos = (int**)malloc((periodo+2)*sizeof(int*));
    for (i = 1; i <= periodo+1; i++) {
      custos[i] = (int*)malloc((maxima+1)*sizeof(int));
    }
    trocas = (int**)malloc((periodo+2)*sizeof(int*));
    for (i = 1; i <= periodo+1; i++) {
      trocas[i] = (int*)malloc((maxima+1)*sizeof(int));
    }
    for (j = 0; j <= maxima; j++) {
      custos[periodo+1][j] = 0;
    }
    /*
      Calcula todas as possibilidades de troca ou manutenção durante o período
      proposto e a idade do destilador de 0 até máxima.

      Para cada (período, idade) verifica se é melhor trocar ou dar manutenção:
        trocar = preço do novo + custo da manutenção - valor da venda + custo acumulado
        manutenção = custo da manutenção + custo acumulado

      Se o valor de trocar for menor ou igual ao valor da manutenção, então é
      recomendado trocar. É adicionado na lista de custos e trocas, os valores:
        custos[período, idade] = valor de trocar
        trocas[período, idade] = True, pq foi trocado
      Caso contrário é recomendado dar manutenção, então é armazenado o
      valor acumulado da manutenção na lista custos.
        custos[(período, idade)] = valor da manuteção
    */
    
    for (i = periodo; i >= 1; i--) {
      for (j = 0; j < maxima; j++) {
        troca = preco + custo[0] - venda[j] + custos[i+1][1];
        manutencao = custo[j] + custos[i+1][j+1];

        if (troca <= manutencao) {
          custos[i][j] = troca;
          trocas[i][j] = 1;
        } else {
          custos[i][j] = manutencao;
          trocas[i][j] = 0;
        }
      }
      /*
        Quando o destilador chega a idade máxima, então tem que trocar
        obrigatoriamente.
      */
      troca = preco + custo[0] - venda[j] + custos[i+1][1];
      custos[i][j] = troca;
      trocas[i][j] = 1;
    }
    printf("%d\n", custos[1][idade]);
    /*
      Verifica os valores dos períodos onde houve troca e monta
      a saída de anos. Os anos de troca são as posições [período, idade],
      onde o período e a idade do destilador foram trocados e que estão
      dentro do período proposto.
    */
    espaco = 1;
    for (i = 1, j = idade; i <= periodo; i++) {
      if (trocas[i][j] == 1) {
        if (!espaco) {
          printf(" ");
        }
        espaco = 0;
        printf("%d", i);
        j = 1;
      } else {
        j++;
      }
    }

    if (espaco) {
      printf("0");
    }
    printf("\n");
    for (i = 1; i <= periodo+1; i++) {
      free(custos[i]);
    }
    free(custos);

    for (i = 1; i <= periodo+1; i++) {
      free(trocas[i]);
    }
    free(trocas);
  }

  return 0;
}